crates/primitives/src/proofs.rs

```diff
@@ -24,63 +25,87 @@ pub const fn adjust_index_for_rlp(i: usize, len: usize) -> usize {
 }
 
 /// Compute a trie root of the collection of rlp encodable items.
-pub fn ordered_trie_root<T: Encodable>(items: &[T]) -> B256 {
+pub fn ordered_trie_root<I, T>(items: I) -> B256
+where
+    T: Encodable,
+    I: IntoIterator<Item = T>,
+{
     ordered_trie_root_with_encoder(items, |item, buf| item.encode(buf))
 }
 
 /// Compute a trie root of the collection of items with a custom encoder.
-pub fn ordered_trie_root_with_encoder<T, F>(items: &[T], mut encode: F) -> B256
+pub fn ordered_trie_root_with_encoder<T, F, I>(items: I, mut encode: F) -> B256
 where
-    F: FnMut(&T, &mut dyn BufMut),
+    I: IntoIterator<Item = T>,
+    F: FnMut(T, &mut dyn BufMut),
 {
     let mut index_buffer = BytesMut::new();
     let mut value_buffer = BytesMut::new();
-
     let mut hb = HashBuilder::default();
-    let items_len = items.len();
-    for i in 0..items_len {
-        let index = adjust_index_for_rlp(i, items_len);
 
+    // Enumerate and encode all keys.
+    let mut remaining = items.into_iter().enumerate().map(|(index, item)| {
         index_buffer.clear();
         index.encode(&mut index_buffer);
-
+        let key = Nibbles::unpack(&index_buffer);
+        (key, item)
+    });
+    let first = remaining.next().into_iter();
+
+    // Since we know all keys will be ordered, except for the key at the 0th index, there is no
+    // need to re-sort the entire iterator.  All we need to do is insert the 0th index item at the
+    // appropriate index (either rlp(0)==0x7f or last).  We can do this by merging two iterators:
+    // one that contains only the first item, and one that contains the remaining items.
```

Comment:
Is this statement correct? Previously we used `adjust_index_for_rlp` to increment the index.

```
/// Adjust the index of an item for rlp encoding.
pub const fn adjust_index_for_rlp(i: usize, len: usize) -> usize {
    if i > 0x7f {
        i
    } else if i == 0x7f || i + 1 == len {
        0
    } else {
        i + 1
    }
}
```

For more than `0x7f` (127) items index becomes `i` not `i+1`